### ISSUE TITLE
There are not 290 distinct states in blackjack

### ISSUE THREAD

**Issue by jxnscott** [jlm429/bettermdptools#26]
```python
self._transform_obs = lambda obs: (
            int(f"{28}{(obs[1] - 2) % 10}") if (obs[0] == 21 and obs[2])
            else int(f"{27}{(obs[1] - 2) % 10}") if (obs[0] == 21 and not obs[2])
            else int(f"{obs[0] + 6}{(obs[1] - 2) % 10}") if obs[2]
            else int(f"{obs[0] - 4}{(obs[1] - 2) % 10}"))
```

In blackjack_wrapper.py, when there is a state tuple `(21, <dealer_card>, 1)`, `else int(f"{obs[0] + 6}{(obs[1] - 2) % 10}") if obs[2]` will never be triggered. In other words, `BJ` and `S21` are indexed by the same tuple. 

```python
>>> def test():
...     observation = (21, 999, 1)
...     if (observation[0] == 21 and observation[2]):
...             print("BJ")
...     elif (observation[2]):
...             print("S21")
...     else:
...             print("No suitable mapping")
... 
>>> test()
BJ
```
In order for there to be 290 distinct states I believe you would need to expand the state space to include the composition of the player's hand.

**Comment by jxnscott** [comment]
Here is a test script I used to check a few things:

```python
import gymnasium as gym
from tqdm.auto import tqdm
base_env = gym.make('Blackjack-v1', render_mode=None)

# Tansform obs function from better mdp tools
transform_obs = lambda obs: (
            int(f"{28}{(obs[1] - 2) % 10}") if (obs[0] == 21 and obs[2])
            else int(f"{27}{(obs[1] - 2) % 10}") if (obs[0] == 21 and not obs[2])
            else int(f"{obs[0] + 6}{(obs[1] - 2) % 10}") if obs[2]
            else int(f"{obs[0] - 4}{(obs[1] - 2) % 10}"))

dealer_cards = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
player_sum = 21

# Find the integer mappings for all state corresponding to the form (21, dealer_card, True)
possible_s21 = []
for dealer_card in dealer_cards:
    obs = (player_sum, dealer_card, True)

    # [270, 271, 272, 273, 274, 275, 276, 277, 278, 279]
    possible_s21.append(int(f"{obs[0] + 6}{(obs[1] - 2) % 10}"))

# Run one million iterations
visited_states = []
for _ in tqdm(range(10_000_000)):
    obs, info = base_env.reset()
    transformed = transform_obs(obs)
    visited_states.append(transformed)

    if transformed in possible_s21:
        print(obs)

    while True:
        action = base_env.action_space.sample()
        obs, reward, terminated, truncated, info = base_env.step(action)

        transformed = transform_obs(obs)
        #if transformed in possible_s21:
            #print(obs)

        # Transform the obs
        visited_states.append(transformed)

        if truncated or terminated:
            if obs[0] > 21:
                print(obs)
                print(transformed)
            break

# Find all unique states
visited_states = list(set(visited_states))
print("Total Unique States Visited: ", len(visited_states))

# Find out if the s21 mapping ever appeared
for s21 in possible_s21:
    if s21 in list(visited_states):
        print("S21 Appeared")
        break

import numpy as np
mappings = list(np.arange(0, 290, 1))
not_visited = []
for value in mappings:
    if value not in visited_states:
        not_visited.append(value)
print(not_visited)

#
# >>> Total Unique States Visited:  280
# >>> S21 Appeared, however
#>>>  [170, 171, 172, 173, 174, 175, 176, 177, 178, 179]
```

> Here is my ed post using this script:
> 
> I am using bettermdptools and the version of gymnasium specified in its setup.py.
> 
> So, I just ran blackjack 1 million times with randomly sampled actions until episode end. The bettermdptools documentation says there are 290 valid unique states. The policy map produced by bettermdptools is also for 290 states.
> 
> However. my hand calculations yielded 280 valid unique states, so I wanted to double check my understanding. I was confused as to why S21 and BJ were being modeled separately by bettermdptools, as both are of the form (21, <dealer_card>, True), which means they are indistinguishable from the perspective of the agent - unless I am misunderstanding what S21 is.
> 
> I recorded every state visited during the 1 million episodes.
> 
> I transformed the visited states into the integer mappings with the bettermdptools transform_obs function. Then I turned that transformed list into a set to get the unique values. I took the length of the set, its 280. Meaning, only 280 unique states were visited.
> 
> I then iterated through every possible permutation of the state set (for player sum 4->21, for dealer card 1->10, for usable ace 0->1). For each permutation, I transformed it into the integer mapping using transform_obs from bettermdptools and tested the mapping to see if it was in my unique state set, created from the 1 million episodes. This test revealed 170-179 was not in my unique state set. These states that were not visited were of the form (11, <dealer_card>, True). If you use transform_obs, the permutations of this form equal 170-179. It makes sense these states would not be visited during the game as its not possible to have a usable ace with a player sum of 11. I am confused as to why this mapping to 170-179 for (11, <dealer_card>, True) exists in the bettermdptools code, as this state is not possible by the rules of the game. 170-179 is supposed to be H21 according to the bettermdptools documentation, which is of the form (21, <dealer_value>, False)
> 
> Going back to my confusion about S21, the mapping values (270-279) for S21 appeared in the unique visited state set, interesting...
> 
> I re-ran through all state permutations and I printed the state tuples for what was mapping to S21, it was tuples of the form (21, <dealer_value>, False). This form is actually H21 based on the documentation.

Essentially, I believe the model is treating one set of states as two separate sets of states (BJ and S21), consequently, the resultant policy and policy mapping are a bit off from the actual game.

**Comment by epoint95** [comment]
Yes just discovered this (you will see this by comparing your model free and model dependent learners). The issue is the transform_obs function is mapping Hard 21s (170-179) to soft 21s (270-279) so we see no Hard 21s (170-179).

**Comment by epoint95** [comment]
Yes just discovered this (you will see this by comparing your model free and model dependent learners). The issue is the transform_obs function is mapping Hard 21s (170-179) to soft 21s (270-279) so we see no Hard 21s (170-179).

**Comment by jlm429** [comment]
fixed/closing - PR #57
